### DISCUSSION TITLE
Inconsistent gray values when using FP3D_CUDA and BP3D_CUDA

### DISCUSSION THREAD

**Discussion by mareikethies** [astra-toolbox/astra-toolbox#410] [Q&A]
Hi,
I am using the FP3D_CUDA to forward project a CT volume. Then, I implement a custom ramp filtering step of the form
```python
filtered_sino = np.zeros_like(sino)
for i in range(sino.shape[0]):
    filtered_slice = np.fft.fft(sino[i, :, :], axis=-1)
    filtered_slice = np.multiply(filtered_slice, ramp_filter)
    filtered_slice = np.fft.ifft(filtered_slice, axis=-1).real
    filtered_sino[i, :, :] = filtered_slice
```
where the number of projections is in the first dimension of the sinogram and the ramp_filter is computed as the fft of the spatial definition of the ramlak filter and repeated to be able to apply it to one 2D projection at once. Finally, I apply the BP3D_CUDA to reconstruct the filtered sinogram. 

Reconstruction works, but I observe that the gray values of the reconstructed reconstructed volume and original volume do not lie in the same range any more. Is this expected? Is there anything I can do to get back comparable gray values?

As my geometry is not perfectly (but almost) circular, I need to use the cone_vec geometry and cannot use FDK_CUDA right away.
Any ideas on this are very appreciated!

**Comment by askorikov** [comment]
Hi! I think this may be related to the additional scaling  you need to apply, like here: https://github.com/astra-toolbox/astra-toolbox/blob/7273498d63451ee75c5389b693fe301f9f04d356/cuda/2d/fbp.cu#L334-L336

**Comment by mareikethies** [comment]
Hi, thanks for your reply! 
I assume M_PI is just pi and dims.iProjAngles is the number of projections in the scan? If that is correct, the above weight does not yield comparable gray values when used as a multiplicative correction factor on the final reconstruction. The file you linked also contains variables fProjectorScale and fReconstructionScale. Can we assume that those are 1?

**Comment by askorikov** [comment]
Now that I'm thinking about it, I guess it's more complicated. Considering you use cone beam geometry, you'll need to do FDK weighting, and the `fProjectorScale` and `fReconstructionScale` factors in your case will probably not equal to 1.0 because they scale with pixel and voxel size. I could suggest looking into Tomosipo library where FDK algorithm is implemented using higher level Python/Pytorch functionality on top of Astra primitives. Specifically, I would look here
https://github.com/ahendriksen/ts_algorithms/blob/9533df05501d5d560a0c2059010dd91b836b85a3/ts_algorithms/fdk.py#L12
and here
https://github.com/ahendriksen/ts_algorithms/blob/9533df05501d5d560a0c2059010dd91b836b85a3/ts_algorithms/fbp.py#L174-L176
